packages/docutils/lib/builder/nav.ts

```diff
@@ -0,0 +1,320 @@
+  // we don't need the `name` prop, since the name is already present in the keypath.
+  const newData: Omit<ParsedNavData, 'name'>[] = [];
+
+  for (const dir of dirs) {
+    const newDataForDir: Omit<ParsedNavData, 'name'>[] = [];
+    const newRefFilepaths = await findRelativeMarkdownFiles(dir, mkDocsDocsDir);
+
+    const header = all ? _.startCase(path.basename(dir)) : DEFAULT_NAV_HEADER;
+
+    const headerItems = itemsForHeader(navData, header);
+
+    // if we found items with this header already, we are going
+    // to replace them all wholesale
+    if (headerItems.length) {
+      shouldWriteMkDocsYml = true;
+      // these are the parts of the keypath of the first item, which will contain the header string.
+      const rootHeaderKeypathParts = _.toPath(_.first(headerItems)!.keypath);
+
+      // this is the keypath up to the header string, inclusive.
+      // we append indices or names to this keypath
+      const rootHeaderKeypath = rootHeaderKeypathParts
+        .slice(0, rootHeaderKeypathParts.indexOf(header) + 1)
+        .join('.');
+
+      for (const [offset, newRefFilepath] of newRefFilepaths.entries()) {
+        const data = headerItems.find((item) => item.fileOrUrl === newRefFilepath);
+        if (data?.name) {
+          newDataForDir.push({
+            keypath: `${rootHeaderKeypath}.${offset}.${data.name}`,
+            fileOrUrl: newRefFilepath,
+          });
+        } else {
+          newDataForDir.push({
+            keypath: `${rootHeaderKeypath}.${offset}`,
+            fileOrUrl: newRefFilepath,
+          });
+        }
+      }
+
+      // look for any differences between what we have and what's in the file
+      if (_.xor(newDataForDir, _.map(headerItems, _.partial(_.omit, _, 'name')))) {
+        newData.push(...newDataForDir);
+        shouldWriteMkDocsYml = true;
+        log.debug('Will write new nav data for header %s', header);
+      } else {
+        log.debug('No changes for header %s', header);
+      }
+    } else {
+      let navOffset = nav.length;
+      for (const [idx, newRefFilepath] of newRefFilepaths.entries()) {
+        newDataForDir.push({
+          keypath: `${navOffset}.${header}.${idx}`,
+          fileOrUrl: newRefFilepath,
+        });
+      }
+      newData.push(...newDataForDir);
+      shouldWriteMkDocsYml = true;
+      log.debug('Will create nav data for header %s', header);
+    }
+  }
+
+  for (const {keypath, fileOrUrl} of newData) {
+    _.set(mkDocsYml, `nav.${keypath}`, fileOrUrl);
+  }
+
+  if (shouldWriteMkDocsYml) {
+    const yaml = stringifyYaml(mkDocsYml);
+    log.debug('Writing to %s:\n%s', mkDocsYmlPath, yaml);
+    await safeWriteFile(mkDocsYmlPath, yaml, true);
+    log.success(
+      'Updated MkDocs navigation config for reference docs; please run "git add -A %s" and commit this change',
+      relMkDocsYmlPath
+    );
+  } else {
+    log.info('No changes needed for MkDocs config at %s', relMkDocsYmlPath);
+  }
```

Comment:
this method is quite long and hard to understand. Maybe it makes sense to extract some code from it?

Reply:
@mykola-mokhnach OK, I've pushed changes based on your review